presto-spi/src/main/java/com/facebook/presto/spi/block/AbstractArrayBlock.java

```diff
@@ -0,0 +1,234 @@
+
+import io.airlift.slice.Slice;
+import io.airlift.slice.Slices;
+
+public abstract class AbstractArrayBlock
+        implements Block
+{
+    protected abstract Block getValues();
+
+    protected abstract Slice getOffsets();
+
+    protected abstract Slice getValueIsNull();
+
+    @Override
+    public BlockEncoding getEncoding()
+    {
+        return new ArrayBlockEncoding(getValues().getEncoding());
+    }
+
+    private int getOffset(int position)
+    {
+        return getOffsets().getInt(position * 4);
+    }
+
+    @Override
+    public Block getRegion(int position, int length)
+    {
+        return getRegion(position, length, false);
+    }
+
+    @Override
+    public Block copyRegion(int position, int length)
+    {
+        return getRegion(position, length, true);
+    }
+
+    private Block getRegion(int position, int length, boolean compact)
+    {
+        int positionCount = getPositionCount();
+        if (position < 0 || length < 0 || position + length > positionCount) {
+            throw new IndexOutOfBoundsException("Invalid position " + position + " in block with " + positionCount + " positions");
+        }
+
+        int startValueOffset = getOffset(position);
+        int endValueOffset = getOffset(position + length);
+        Block newValues;
+        Slice newOffsets;
+        Slice newValueIsNull;
+        if (compact) {
+            newValues = getValues().copyRegion(startValueOffset, endValueOffset - startValueOffset);
+            int[] newOffsetsArray = new int[length + 1];
+            for (int i = 0; i <= length; i++) {
+                newOffsetsArray[i] = getOffsets().getInt((position + i) * 4) - startValueOffset;
+            }
+            newOffsets = Slices.wrappedIntArray(newOffsetsArray);
+            newValueIsNull = Slices.copyOf(getValueIsNull(), position, length);
+            // In the case of compact, one could argue that is is beneficial to always set startValueOffset to 0, and adjust newOffsets accordingly.
+            // That would result in a unique in-memory representation for ArrayBlock when compacted.
+        }
+        else {
+            if (position == 0 && length == positionCount) {
+                // It is incorrect to pull up this `if` because child blocks may be compact-able when this if condition is satisfied
+                return this;
+            }
+            else {
+                newValues = getValues();
+                newOffsets = getOffsets().slice(position * 4, (length + 1) * 4);
+                newValueIsNull = getValueIsNull().slice(position, length);
+            }
+        }
+        return new ArrayBlock(newValues, newOffsets, newValueIsNull);
+    }
+
+    @Override
+    public int getLength(int position)
+    {
+        return getOffset(position + 1) - getOffset(position);
+    }
+
+    @Override
+    public int getSizeInBytes()
+    {
+        return Integer.BYTES + getValues().getSizeInBytes() + getOffsets().length() + getValueIsNull().length();
```

Comment:
Why + `Integer.BYTES`? `ArrayBlock` doesn't seem to have anything that's the size of an int, and `ArrayBlockBuilder` has more